### PR TITLE
pipe: add uv_pipe and uv_socketpair to the API (v1.x)

### PR DESCRIPTION
**Pull request by vtjnash**
Enhanced equivalents of `pipe` and `socketpair` for cross-platform. Ports #1498 back to v1.x, since I'm getting really tired of rebasing it, and the API on v1.x is only slightly worse for implementing it (`uv_file` vs `uv_os_fd_t`).

CI: https://ci.nodejs.org/job/libuv-test-commit/1986/

### CODE REVIEW

File: src/unix/pipe.c [hunk 1]
```diff
@@ -379,3 +379,51 @@ int uv_pipe_chmod(uv_pipe_t* handle, int mode) {
...
+  if (write_flags & UV_NONBLOCK_PIPE)
+    if ((err = uv__nonblock(fds[1], 1)))
+      goto fail;
+
+  return 0;
+
+fail:
+  uv__close(fds[0]);
+  uv__close(fds[1]);
+  return err;
```
Comment (bnoordhuis) [inline src/unix/pipe.c:421]: It'd be better to not clobber `fds` on error. Same suggestion applies to `uv_socketpair()`.

File: test/test-ping-pong.c [hunk 2]
```diff
@@ -148,8 +160,44 @@ static void pinger_read_cb(uv_stream_t* stream,
...
+  }
+
+  /* Echo back */
+  for (i = 0; i < nread; i++) {
+    if (buf->base[i] == 'I')
+      buf->base[i] = 'O';
+  }
+
+  writebuf = uv_buf_init(buf->base, nread);
+  req = malloc(sizeof(*req));
```
Comment (bnoordhuis) [inline test/test-ping-pong.c:188]:
```suggestion
  req = malloc(sizeof(*req));
  ASSERT_NOT_NULL(req);
```

File: test/test-ping-pong.c [hunk 3]
```diff
@@ -148,8 +160,44 @@ static void pinger_read_cb(uv_stream_t* stream,
...
+  if (uv_write(req,
+               stream,
+               &writebuf,
+               1,
+               pinger_after_write)) {
+    FATAL("uv_write failed");
+  }
```
Comment (bnoordhuis) [inline test/test-ping-pong.c:195]: I'd use `ASSERT_EQ(0, uv_write(...))`, that prints the status code.

File: test/test-ping-pong.c [hunk 4]
```diff
@@ -252,6 +303,75 @@ static void pipe_pinger_new(int vectored_writes) {
...
+  pinger = (pinger_t*) malloc(sizeof(*pinger));
+  ASSERT(pinger != NULL);
+  pinger->vectored_writes = vectored_writes;
+  pinger->state = 0;
+  pinger->pongs = 0;
+  pinger->pong = PONG;
+
+  /* Try to make a socketpair and do NUM_PINGS ping-pongs. */
+  (void)uv_default_loop(); /* ensure WSAStartup has been performed */
+  ASSERT(0 == uv_socketpair(SOCK_STREAM, 0, fds, UV_NONBLOCK_PIPE, UV_NONBLOCK_PIPE));
```
Comment (bnoordhuis) [inline test/test-ping-pong.c:320]: `ASSERT_EQ(0, ...)` here and below. :-)

File: test/test-ping-pong.c [hunk 5]
```diff
@@ -252,6 +303,75 @@ static void pipe_pinger_new(int vectored_writes) {
...
+  /* On Windows, this is actually a UV_TCP, but libuv doesn't detect that. */
+  ASSERT(uv_guess_handle((uv_file) fds[0]) == UV_NAMED_PIPE);
+  ASSERT(uv_guess_handle((uv_file) fds[1]) == UV_NAMED_PIPE);
+#endif
+
+  ASSERT(0 == uv_tcp_init(uv_default_loop(), &pinger->stream.tcp));
+  pinger->stream.pipe.data = pinger;
+  ASSERT(0 == uv_tcp_open(&pinger->stream.tcp, fds[1]));
+
+  ponger = (uv_tcp_t*) malloc(sizeof(uv_tcp_t));
```
Comment (bnoordhuis) [inline test/test-ping-pong.c:331]:
```suggestion
  ponger = malloc(sizeof(*ponger));
```

File: test/test-ping-pong.c [hunk 6]
```diff
@@ -252,6 +303,75 @@ static void pipe_pinger_new(int vectored_writes) {
...
+  ASSERT(ponger != NULL);
```
Comment (bnoordhuis) [inline test/test-ping-pong.c:332]:
```suggestion
  ASSERT_NOT_NULL(ponger);
```

File: test/test-ping-pong.c [hunk 7]
```diff
@@ -252,6 +303,75 @@ static void pipe_pinger_new(int vectored_writes) {
...
+  ponger->data = NULL;
+  ASSERT(0 == uv_tcp_init(uv_default_loop(), ponger));
+  ASSERT(0 == uv_tcp_open(ponger, fds[0]));
+
+  pinger_write_ping(pinger);
+
+  uv_read_start((uv_stream_t*) &pinger->stream.tcp, alloc_cb, pinger_read_cb);
+  uv_read_start((uv_stream_t*) ponger, alloc_cb, ponger_read_cb);
```
Comment (bnoordhuis) [inline test/test-ping-pong.c:340]: Can you `ASSERT_EQ(0, uv_read_start(...));` here?

Similar suggestions apply to the function below.

File: src/unix/pipe.c [hunk 8]
```diff
@@ -379,3 +379,51 @@ int uv_pipe_chmod(uv_pipe_t* handle, int mode) {
 
   return r != -1 ? 0 : UV__ERR(errno);
 }
+
+
+int uv_pipe(uv_os_fd_t fds[2], int read_flags, int write_flags) {
+  int err;
+#if defined(__FreeBSD__) || defined(__linux__)
+  int flags = O_CLOEXEC;
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/unix/pipe.c:387]: I'd assume `O_CLOEXEC` is always a constant, so this is permitted by the style guide which says that variable declarations must go before code (constant initializers essentially count as declarations and not code in C89).

**Comment by vtjnash** [comment]
CI is showing problems with Win2012, which is a bit inconvenient: https://ci.nodejs.org/job/libuv-test-commit/2025/

**Comment by vtjnash** [comment]
CI ✅: https://ci.nodejs.org/job/libuv-test-commit/2030/ I just forgot how old 2012 was and was using a newer API.

**Comment by vtjnash** [comment]
Bump. Please review!

File: src/unix/tcp.c [hunk 9]
```diff
@@ -459,3 +459,49 @@ int uv_tcp_simultaneous_accepts(uv_tcp_t* handle, int enable) {
...
+
+int uv_socketpair(int type, int protocol, uv_os_sock_t fds[2], int flags0, int flags1) {
+  uv_os_sock_t temp[2];
+  int err;
+#if defined(__FreeBSD__) || defined(__linux__)
+  int flags;
+
+  flags = type | SOCK_CLOEXEC;
+  if ((flags0 & UV_NONBLOCK_PIPE) && (flags1 & UV_NONBLOCK_PIPE))
+    flags |= UV_FS_O_NONBLOCK;
```
Comment (santigimeno) [inline src/unix/tcp.c:472]: Any reason not using `SOCK_NONBLOCK` directly?

**Review by santigimeno — APPROVED** [review]
Unix changes LGTM with a couple of comments. Thanks

**Comment by vtjnash** [comment]
Thanks, those were rebase mistakes that snuck in. I would have liked at least one more reviewer, but would like to merge this soon regardless.

CI: https://ci.nodejs.org/job/libuv-test-commit/2086/

**Pull request merged**